storage/remote/codec.go

```diff
@@ -183,6 +164,154 @@ func FromQueryResult(res *prompb.QueryResult) storage.SeriesSet {
 	}
 }
 
+// StreamChunkedReadResponses iteraties over series, build chunks and streams those to caller.
+// TODO(bwplotka): Encode only what's needed. Fetch the encoded series from blocks instead of rencoding everything.
+func StreamChunkedReadResponses(
+	stream io.Writer,
+	queryIndex int64,
+	ss storage.SeriesSet,
+	sortedExternalLabels []prompb.Label,
+	maxChunksInFrame int,
+) error {
+	var (
+		chks = make([]prompb.Chunk, 0, maxChunksInFrame)
+		err  error
+	)
+
+	for ss.Next() {
+		series := ss.At()
+		iter := series.Iterator()
+		lbls := MergeLabels(labelsToLabelsProto(series.Labels()), sortedExternalLabels)
+
+		// TODO(bwplotka): We send each series in separate frame no matter what. Even if series has only one sample.
```

Comment:
@brian-brazil @tomwilkie I think this is quite major optimization we can make, but adds extra complexity to the caller. I think it's worth it though.

Reply:
I'm not sure it'll make much difference, series ordering should be preserved though.

Reply:
Yeah this is fine to start, @bwplotka.  Could rephrase for clarity "A most one series per frame; series may be split over multiple frames according to maxChunksInFrame"